### PR TITLE
Rest API Query Syntax Enhancements

### PR DESCRIPTION
**Pull request by peterbroadhurst**
Proposal to improve the rich query semantics of FireFly, to support "starts with" and "ends with", and add extra flexibility/consistency.

### CODE REVIEW

File: text/0000-rest-query-syntax.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+Modifiers can appear before the operator, to change its
+behavior.
+
+| Modifier | Description                        |
+|----------|------------------------------------|
+| `!`    * | Not - negates the match            |
+| `:`    * | Case insensitive                   |
+
+> [*] - these characters should technically be encoded in URL queries,
+>       but in practice have been tested to function fine without encoding.
```
Comment (awrichar) [inline text/0000-rest-query-syntax.md:91]: If we're going by RFC3986, query strings are actually pretty permissive. There are a few chars not allowed, like `#` or `[]`. But everything here seems fine.
```
   unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
   sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="
   pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
   query         = *( pchar / "/" / "?" )
```

**Comment by awrichar** [comment]
Looks really good to me. Feels like a nice set of params for the vast majority of cases, and addresses my long gripe against our `^` and `@` usage :)

File: text/0000-rest-query-syntax.md (hunk 1 above)
Comment (peterbroadhurst) [reply text/0000-rest-query-syntax.md:91]: ... in the PR docs, I'll leave this bit out

**Comment by nguyer** [comment]
This looks great to me. I realize it's a breaking change. Normally we would need to plan a way to mitigate breakage and deprecate the old behavior, but we're still pre-1.0 at this point, so I think we can just make this change in place. That's a convenience we won't be able to afford going forward thoguh.

**Comment by peterbroadhurst** [comment]
@awrichar @nguyer - made one small extension as I was closing out the implementation, and realized we're missing a trick of allowing people to distinguish between an empty value and nil in queries

File: text/0000-rest-query-syntax.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+### Modifiers
+
+Modifiers can appear before the operator, to change its
+behavior.
+
+| Modifier | Description                        |
+|----------|------------------------------------|
+| `!`      | Not - negates the match            |
+| `:`      | Case insensitive                   |
+| `;`      | Treat empty match string as null   |
```
Comment (peterbroadhurst) [inline text/0000-rest-query-syntax.md:89]: This is the modifier to allow specifying as null (found this much clearer in the end than a new operator)

File: text/0000-rest-query-syntax.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+| `=cat`       | Equals "cat" (same)                        |
+| `!=cat`      | Not equal to "cat"                         |
+| `:=cat`      | Equal to "CAT", "cat", "CaT etc.           |
+| `!:cat`      | Not equal to "CAT", "cat", "CaT etc.       |
+| `=!cat`      | Equal to "!cat" (! is after operator)      |
+| `^cats/`     | Starts with "cats/"                        |
+| `$_cat`      | Ends with with "_cat"                      |
+| `!:^cats/`   | Does not start with "cats/", "CATs/" etc.  |
+| `!$-cat`     | Does not end with "-cat"                   |
+| `;=`         | Is null                                    |
```
Comment (peterbroadhurst) [inline text/0000-rest-query-syntax.md:103]: Example here of using it, and below of negating it.

File: text/0000-rest-query-syntax.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+| `!;=`        | Is not null                                |
```
Comment (peterbroadhurst) [inline text/0000-rest-query-syntax.md:104]: Examples here of using it

File: text/0000-rest-query-syntax.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+### Modifiers
+
+Modifiers can appear before the operator, to change its
+behavior.
+
+| Modifier | Description                        |
+|----------|------------------------------------|
+| `!`      | Not - negates the match            |
+| `:`      | Case insensitive                   |
+| `?`      | Treat empty match string as null   |
```
Comment (peterbroadhurst) [inline text/0000-rest-query-syntax.md:89]: This is the modifier to allow specifying as null (found this much clearer in the end than a new operator)

File: text/0000-rest-query-syntax.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+| `!=cat`      | Not equal to "cat"                         |
+| `:=cat`      | Equal to "CAT", "cat", "CaT etc.           |
+| `!:cat`      | Not equal to "CAT", "cat", "CaT etc.       |
+| `=!cat`      | Equal to "!cat" (! is after operator)      |
+| `^cats/`     | Starts with "cats/"                        |
+| `$_cat`      | Ends with with "_cat"                      |
+| `!:^cats/`   | Does not start with "cats/", "CATs/" etc.  |
+| `!$-cat`     | Does not end with "-cat"                   |
+| `?=`         | Is null                                    |
+| `!?=`        | Is not null                                |
```
Comment (peterbroadhurst) [inline text/0000-rest-query-syntax.md:104]: Examples here of using it

File: text/0000-rest-query-syntax.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+| `!=cat`      | Not equal to "cat"                         |
+| `:=cat`      | Equal to "CAT", "cat", "CaT etc.           |
+| `!:cat`      | Not equal to "CAT", "cat", "CaT etc.       |
+| `=!cat`      | Equal to "!cat" (! is after operator)      |
+| `^cats/`     | Starts with "cats/"                        |
+| `$_cat`      | Ends with with "_cat"                      |
+| `!:^cats/`   | Does not start with "cats/", "CATs/" etc.  |
+| `!$-cat`     | Does not end with "-cat"                   |
+| `?`          | Is null                                    |
+| `!?`         | Is not null                                |
```
Comment (peterbroadhurst) [inline text/0000-rest-query-syntax.md:104]: Examples here of using it

**Comment by nguyer** [comment]
Moving this FIR into the Final Comment Period, set to close on 2022-01-17

**Review by nguyer — APPROVED** [review]

**Pull request merged**